gwcs/selector.py

```diff
@@ -1,87 +1,168 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
 from __future__ import division, print_function
 
 import copy
 import numpy as np
 import warnings
 
+from astropy.extern import six
 from astropy.modeling.core import Model
-from astropy.utils.misc import isiterable
-
+from astropy.modeling.parameters import Parameter
 
+from . import region
 from .util import RegionError
 
 
-class SelectorModel(Model):
+def _toindex(value):
+    """
+    Convert value to an int or an int array.
+
+    Input coordinates should be turned into integers
+    corresponding to the center of the pixel.
+    They are used to index the mask.
+
+    Examples
+    --------
+    >>> _toindex(np.array([-0.5, 0.49999]))
+    array([0, 0])
+    >>> _toindex(np.array([0.5, 1.49999]))
+    array([1, 1])
+    >>> _toindex(np.array([1.5, 2.49999]))
+    array([2, 2])
+    """
+    indx = np.empty(value.shape, dtype=np.int32)
+    indx = np.floor(value + 0.5, out=indx)
+    return indx
+
+
+class SelectorMask(Model):
+
     """
+    A mask model to be used with the `~gwcs.selector.RegionsSelector` transform.
+
+    For an IFU observation, the values of the mask
+    correspond to the region (slice)  label.
+
     Parameters
     ----------
-    labels : a list of strings or objects
-    transforms : a list of transforms
-        transforms match labels
+    mask : ndarray
+        An array of integers or strings where the values
+        correspond to a transform label in `~gwcs.selector.RegionsSelector` model.
+        If a transform is not defined the value shoul dbe set to 0 or " ".
     """
-    _param_names = ()
+
+    inputs = ('x', 'y')
+    outputs = ('z')
 
     linear = False
     fittable = False
 
-    def __init__(self, labels, transforms):
-        # labels should be unique
-        if (np.unique(labels)).size != np.asarray(labels).size:
-            raise ValueError("Labels should be unique.")
-        # every label should have a corresponding transform
-        if len(labels) != len(transforms):
-            raise ValueError("Expected labels and transforms to be of equal length.")
-        if " " in labels or 0 in labels:
-            raise ValueError('"0" and " " are not allowed as labels.')
-        self._selector = dict(zip(labels, transforms))
-        super(SelectorModel, self).__init__(n_models=1)
+    def __init__(self, mask):
+        if mask.dtype.type is not np.string_:
+            self._mask = np.asarray(mask, dtype=np.int)
+        else:
+            self._mask = mask
+        if mask.dtype.type is np.string_:
+            self._no_transform_value = ""
+        else:
+            self._no_transform_value = 0
+        super(SelectorMask, self).__init__()
 
```

Comment:
@embray @mdboom 

This model has one parameter `mask` which is an array of type `int` or `str`. Because of the type of the array I cannot write it as a typical `modeling.Model` which converts all parameters to float. That's why it's written the way it is.
@embray Is there a better way to implement it?

@mdboom When testing the schema for this model, the `AsdfFile.writeto` method  gives me

```
ValueError: Inconsistent data column lengths: set([])
```
If I write `SelectorMask.mask` as `modeling.Parameter` then this error goes away (but it is now of type float, which I am trying to avoid).  Is pyasdf relying on any `modeling` specific methods or attributes to write out the model? The schema I am using looks like this:
```
%YAML 1.1
---
$schema: "http://stsci.edu/schemas/yaml-schema/draft-01"
id: "http://stsci.edu/schemas/asdf/0.1.0/transform/selector_mask"
tag: "tag:stsci.edu:asdf/0.1.0/transform/selector_mask"
title: >
  Represents a SelectorMask.

allOf:
  - $ref: transform
  - type: object
    properties:
      mask:
        type: !core/ndarray
    required: [mask]
```